### ISSUE TITLE
please detect duplicate shares and don't submit them to the pool

### ISSUE THREAD

**Issue by aleqx** [xmrig/xmrig-proxy#320]
When multiple rigs are connected to the proxy, it can happen that there are duplicate shares if 2 or more miners happen to start with the same nonce and search space. This results in duplicate shares, which the pool will reject. The bigger problem is that a LOT of pools are badly configured and they ban the miner IP address if it sends 2 duplicate shares in a short time interval. See screenshot below. This is an even bigger problem for larger farms with many workers.

The proxy CAN detect if a share is a duplicate. Could you please implement a duplicate share detection and then don't submit duplicates to the pool?

The best would be for xmrig-proxy and xmrig to talk to each other so that each worker connected to the proxy gets a different nonce search space so that they never overlap.

![](https://i.gyazo.com/27ceb026fa5675e9950b5181c9dc709d.png)

**Comment by xmrig** [comment]
Actually both proxy and miner detects duplicate jobs from pool and this is only known reason for duplicated shares https://github.com/xmrig/xmrig-proxy/blob/master/src/common/net/Client.cpp#L379

Please rebuild miner with debug log `-DWITH_DEBUG_LOG=ON`, after that proxy will show network log, it helps understand what exactly happen.
Thank you.

**Comment by aleqx** [comment]
I think you misunderstood. I'm not talking about the miner or proxy detecting the pool response about duplicate shares. I'm talking about the proxy detecting duplicate shares BEFORE they are sent to the pool, i.e. the pool would never receive a duplicate share.

In the screenshot above, the proxy sends duplicate shares to the pool. However, it should be able to detect that a share is a duplicate and not send it to the pool.

**Comment by xmrig** [comment]
Yep, it possible add checks for this case too, but in normal workflow such error should never happen, so please provide the log and/or provide more information. My primary interest is to fix the root cause on miner side, not just filter duplicates.
Thank you.

**Comment by aleqx** [comment]
I did explain in the 1st post. When you connect many workers to the proxy then it becomes possible for 2 or more of them to find the exact same share (nonce). This is called share collision. You can't prevent it unless you make that each worker uses different nonce search space.

For now, your proxy receives all shares and can see if there is a collision (duplicate), and should drop duplicates -- don't send both to the pool because the pool will reject one of them, and many pools also ban or disconnect miners that do.

I'd say this is easier to implement as a first stage.

You could then spend more time to implement a stratum protocol extension for new job JSON-RPC messages, to tell each miner which nonce search space to use in order to avoid share collisions.

**Comment by xmrig** [comment]
> each worker uses different nonce search space.

But it already done, by using nicehash method, nonce reduced to 3 bytes and not overlapping.

**Comment by aleqx** [comment]
But that requires that I use xmrig for the miner software, whereas I use CryptoDredge (because it's much faster than xmrig on Nvidia).

The improvement I proposed, for xmrig-proxy to drop duplicates before sending them to the pool, is universal (works for any miner software).

**Comment by xmrig** [comment]
nicehash option for xmrig miners doesn't matter for recent versions, it enabled automatically if need.

If miner software not support nicehash (why? it very simple thing) you have 2 options:
1. Switch proxy to simple mode, but it disable any benefit of reducing connection count, 16 miners will become 16 pool connections, instead of 1.
2. Don't use proxy at all, 16 is small number for any pool.

Ignoring duplicates is not a option, it cut your effective hashrate.

**Comment by aleqx** [comment]
> Ignoring duplicates is not a option, it cut your effective hashrate.

That makes no sense. Please read my 1st post again. The pool rejects duplicates anyway (so you lose hashrate regardless) but if you send duplicates to the pool then many pools disconnect and ban the miner for 10-15 minutes, which is worse.

I posted a full screenshot of this happening. It wouldn't happen if xmrig-proxy dropped the duplicate shares (do not send them to the pool).

> Don't use proxy at all, 16 is small number for any pool.

Not true. I've met pools which were banning the miner due to too many connections. Also, many pools are badly configured in terms of vardiff. Using xmrig-proxy allows me to control the diff to miners.

> If miner software not support nicehash (why? it very simple thing)

You can talk to the CryptoDredge developer about that if you wish. https://github.com/technobyl/CryptoDredge/issues
